### ISSUE TITLE
Problem with Ajax listener method with PF6.0

### ISSUE THREAD

**Issue by ybendek** [primefaces/primefaces#1594]
Hello

I have been using PF in a project and we are starting to migrate from PF5.3 to PF6.0. However, now from any Ajax call using the listener we are having the same exception.
`Grave: javax.el.MethodNotFoundException: Return type 'void' of method 'onToggleClientData' in 'class com.unitas.view.reports.AttendanceActReportMngBean' does not match 'class java.lang.Object'
    at de.odysseus.el.tree.impl.ast.AstProperty.findMethod(AstProperty.java:174)
    at de.odysseus.el.tree.impl.ast.AstProperty.invoke(AstProperty.java:203)
    at de.odysseus.el.tree.impl.ast.AstEval.invoke(AstEval.java:71)
    at de.odysseus.el.TreeMethodExpression.invoke(TreeMethodExpression.java:132)
    at org.apache.myfaces.view.facelets.el.ContextAwareTagMethodExpression.invoke(ContextAwareTagMethodExpression.java:96)
    at org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl.processCustomArgListener(AjaxBehaviorListenerImpl.java:107)
    at org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl.processArgListener(AjaxBehaviorListenerImpl.java:82)
    at org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl.processAjaxBehavior(AjaxBehaviorListenerImpl.java:63)
    at org.primefaces.event.AbstractAjaxBehaviorEvent.processListener(AbstractAjaxBehaviorEvent.java:38)
    at javax.faces.component.behavior.BehaviorBase.broadcast(BehaviorBase.java:70)
    at javax.faces.component.UIComponentBase.broadcast(UIComponentBase.java:415)
    at javax.faces.component.UIViewRoot._broadcastAll(UIViewRoot.java:1172)
    at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:354)
    at javax.faces.component.UIViewRoot._process(UIViewRoot.java:1658)
    at javax.faces.component.UIViewRoot.processDecodes(UIViewRoot.java:869)
    at org.apache.myfaces.lifecycle.ApplyRequestValuesExecutor.execute(ApplyRequestValuesExecutor.java:42)
    at org.apache.myfaces.lifecycle.LifecycleImpl.executePhase(LifecycleImpl.java:196)
    at org.apache.myfaces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:143)
    at javax.faces.webapp.FacesServlet.service(FacesServlet.java:198)
    at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1550)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:343)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:217)
    at org.primefaces.webapp.filter.FileUploadFilter.doFilter(FileUploadFilter.java:100)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:217)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:279)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:175)
    at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:655)
    at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:595)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:161)
    at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:331)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:231)
    at com.sun.enterprise.v3.services.impl.ContainerMapper$AdapterCallable.call(ContainerMapper.java:317)
    at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:195)
    at com.sun.grizzly.http.ProcessorTask.invokeAdapter(ProcessorTask.java:860)
    at com.sun.grizzly.http.ProcessorTask.doProcess(ProcessorTask.java:757)
    at com.sun.grizzly.http.ProcessorTask.process(ProcessorTask.java:1056)
    at com.sun.grizzly.http.DefaultProtocolFilter.execute(DefaultProtocolFilter.java:229)
    at com.sun.grizzly.DefaultProtocolChain.executeProtocolFilter(DefaultProtocolChain.java:137)
    at com.sun.grizzly.DefaultProtocolChain.execute(DefaultProtocolChain.java:104)
    at com.sun.grizzly.DefaultProtocolChain.execute(DefaultProtocolChain.java:90)
    at com.sun.grizzly.http.HttpProtocolChain.execute(HttpProtocolChain.java:79)
    at com.sun.grizzly.ProtocolChainContextTask.doCall(ProtocolChainContextTask.java:54)
    at com.sun.grizzly.SelectionKeyContextTask.call(SelectionKeyContextTask.java:59)
    at com.sun.grizzly.ContextTask.run(ContextTask.java:71)
    at com.sun.grizzly.util.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:532)
    at com.sun.grizzly.util.AbstractThreadPool$Worker.run(AbstractThreadPool.java:513)
    at java.lang.Thread.run(Thread.java:745)`

Basically now PF6.0 is requesting that any listener method implementation for an ajax call must return something, where is this documented, because this was not a requirenment before. **And we have lot of calls, for example used with events like  toggles column implementations or tab change listeners....**

**Comment by tandraschko** [comment]
we need a example, otherwise we will close it as invalid.

**Comment by ybendek** [comment]
Sure, a very simple example, here using tabView component:

`
<h:form id="formCmp" >
 <p:tabView>
  <p:ajax event="tabChange" listener="#{testBean.onChangeTab}" />
  <p:tab title="tab1"></p:tab>
  <p:tab title="tab2"></p:tab>
  <p:tab title="tab3"></p:tab>
 </p:tabView>
</h:form> 
`

`
public void onChangeTab(TabChangeEvent event){
        System.out.println(event.getTab().getTitle());        
    }
`

In this case the listener method onChangeTab is never called, instead you have this exception:

`Grave: Return type 'void' of method 'onChangeTab' in 'class com.unitas.view.TestMngBean' does not match 'class java.lang.Object'
javax.el.MethodNotFoundException: Return type 'void' of method 'onChangeTab' in 'class com.unitas.view.TestMngBean' does not match 'class java.lang.Object'
    at de.odysseus.el.tree.impl.ast.AstProperty.findMethod(AstProperty.java:174)
    at de.odysseus.el.tree.impl.ast.AstProperty.invoke(AstProperty.java:203)
    at de.odysseus.el.tree.impl.ast.AstEval.invoke(AstEval.java:71)
    at de.odysseus.el.TreeMethodExpression.invoke(TreeMethodExpression.java:132)
    at org.apache.myfaces.view.facelets.el.ContextAwareTagMethodExpression.invoke(ContextAwareTagMethodExpression.java:96)
    at org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl.processCustomArgListener(AjaxBehaviorListenerImpl.java:107)
    at org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl.processArgListener(AjaxBehaviorListenerImpl.java:82)
    at org.primefaces.behavior.ajax.AjaxBehaviorListenerImpl.processAjaxBehavior(AjaxBehaviorListenerImpl.java:63)
    at org.primefaces.event.AbstractAjaxBehaviorEvent.processListener(AbstractAjaxBehaviorEvent.java:38)
    at javax.faces.component.behavior.BehaviorBase.broadcast(BehaviorBase.java:70)
    at javax.faces.component.UIComponentBase.broadcast(UIComponentBase.java:415)
    at org.primefaces.component.api.UITabPanel.broadcast(UITabPanel.java:1259)
    at javax.faces.component.UIViewRoot._broadcastAll(UIViewRoot.java:1172)
    at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:365)
    at javax.faces.component.UIViewRoot._process(UIViewRoot.java:1658)
    at javax.faces.component.UIViewRoot.processApplication(UIViewRoot.java:862)
    at org.apache.myfaces.lifecycle.InvokeApplicationExecutor.execute(InvokeApplicationExecutor.java:42)
    at org.apache.myfaces.lifecycle.LifecycleImpl.executePhase(LifecycleImpl.java:196)
    at org.apache.myfaces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:143)
    at javax.faces.webapp.FacesServlet.service(FacesServlet.java:198)
    at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1550)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:343)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:217)
    at org.primefaces.webapp.filter.FileUploadFilter.doFilter(FileUploadFilter.java:100)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:217)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:279)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:175)
    at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:655)
    at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:595)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:161)
    at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:331)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:231)
    at com.sun.enterprise.v3.services.impl.ContainerMapper$AdapterCallable.call(ContainerMapper.java:317)
    at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:195)
    at com.sun.grizzly.http.ProcessorTask.invokeAdapter(ProcessorTask.java:860)
    at com.sun.grizzly.http.ProcessorTask.doProcess(ProcessorTask.java:757)
    at com.sun.grizzly.http.ProcessorTask.process(ProcessorTask.java:1056)
    at com.sun.grizzly.http.DefaultProtocolFilter.execute(DefaultProtocolFilter.java:229)
    at com.sun.grizzly.DefaultProtocolChain.executeProtocolFilter(DefaultProtocolChain.java:137)
    at com.sun.grizzly.DefaultProtocolChain.execute(DefaultProtocolChain.java:104)
    at com.sun.grizzly.DefaultProtocolChain.execute(DefaultProtocolChain.java:90)
    at com.sun.grizzly.http.HttpProtocolChain.execute(HttpProtocolChain.java:79)
    at com.sun.grizzly.ProtocolChainContextTask.doCall(ProtocolChainContextTask.java:54)
    at com.sun.grizzly.SelectionKeyContextTask.call(SelectionKeyContextTask.java:59)
    at com.sun.grizzly.ContextTask.run(ContextTask.java:71)
    at com.sun.grizzly.util.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:532)
    at com.sun.grizzly.util.AbstractThreadPool$Worker.run(AbstractThreadPool.java:513)
    at java.lang.Thread.run(Thread.java:745)`

If you change the method to:

`public Object onChangeTab(TabChangeEvent event){
        System.out.println(event.getTab().getTitle());        
   return null
    }`

It works, but that is not the expected behavior or at least the documented... and that return was not required before with 5.3

**Comment by tandraschko** [comment]
Workaround: 
http://juel.sourceforge.net/guide/advanced/index.html - Enabling/Disabling ignoring of expected return type

**Comment by tandraschko** [comment]
also fixed in trunk, please give it a try.

**Comment by ybendek** [comment]
@tandraschko but is this including others events? I found exactly the same issue with others event components, for example with "toggle" event in p:columnToggler.

When 6.1 will be delivered (aprox)?

**Comment by tandraschko** [comment]
for every event

probably autumn/winter this year

**Comment by tandraschko** [comment]
This occurs for every event.
The problem is actually that JUEL is the only implementation which handles this case correctly. As we don't test with JUEL per default, we don't hit this bug. But as i posted, the workaround/juel setting works fine.

probably winter/autumn this year.
